File: src/repository/AirportDao.java
package repository;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Consumer;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.EntityTransaction;
import javax.persistence.Persistence;
import javax.persistence.Query;

import model.Airport;

public class AirportDao implements Repository<Airport, Integer>{
	
	private EntityManagerFactory emf = Persistence.createEntityManagerFactory(
			"jdbc:sqlserver://localhost/Alibaba.odb;integratedSecurity=true;");
	private EntityManager em=emf.createEntityManager(); 

	@Override
	public Airport findbyId(Integer id) {
		try {
			return em.find(Airport.class, id);
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<Airport> findByIDs(Collection<Integer> ids) {
		List<Airport> airports = new ArrayList<Airport>();
		try {
			for(int id : ids) {
				airports.add(em.find(Airport.class, id));
			}
			return airports;
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<Airport> findAll() {
		Query q = em.createQuery("select * from Airport");
		return q.getResultList();
	}

	@Override
	public Boolean deleteByID(Integer id) {
		try {
			executeInsideTransaction(entityManager -> em.remove(em.find(Airport.class, id)));
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public Boolean DeleteByIDs(Collection<Integer> ids) {
		try {
			for(int id : ids) {
				executeInsideTransaction(entityManager -> em.remove(em.find(Airport.class, id)));
			}
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public Airport save(Airport E) {
		try {
			executeInsideTransaction(entityManager -> entityManager.persist(E));
			return em.find(Airport.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		} catch(Exception e) {
			executeInsideTransaction(entityManager -> entityManager.merge(E));
			return em.find(Airport.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		}
	}
	
	private void executeInsideTransaction(Consumer<EntityManager> action) {
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            action.accept(em);
            tx.commit(); 
        }
        catch (RuntimeException e) {
            tx.rollback();
            throw e;
        }
    }

}

File: repository/AirportDao.py
from Repository import RepositoryInterface
from Connection import getConn
from Airport import Airport
class AirportDao(RepositoryInterface):
    def findById(id):
        try:
            c = getConn()
            c.execute('select * from Airport where code=' + str(id))
            i = c.fetchall()
            return Airport(i[0][0], i[0][1], i[0][2])
        except:
            print("There was an error")
            return None
    def findByIds(ids):
        try:
            a = []
            c = getConn()
            for id in ids:
                c.execute('select * from Airport where code=' + str(id))
                i = c.fetchall()
                a.append(Airport(i[0][0], i[0][1], i[0][2]))
            return a
        except:
            print("There was an error")
            return None
    def findAll():
        try:
            a = []
            c=getConn()
            c.execute('select * from Airport')    
            for i in c:
                a.append(Airport(i[0], i[1], i[2]))
            return a
        except:
            print("There was an error")
            return None
    def deleteById(id):
        try:
            c=getConn()
            c.execute('delete from Airport where code=' + str(id))
            c.commit()
            return True
        except:
            return False
    def deleteByIds(ids):
        try:
            c=getConn()
            for id in ids:
                   c.execute('delete from Airport where code=' + str(id))
                   c.commit()
            return True
        except:
            return False
    def save(entity):
        try:
            c=getConn()
            c.execute('select * from Airport where code=' + str(entity.code))
            i = c.fetchall()
            if len(i) == 0:
                c.execute("insert into Airport values('" + str(entity.name)+"',"+ str(entity.cityId)+')')
                c.commit()
            else :
                c.execute("update Airport set name='" +str(entity.name)+"', cityId="+ str(entity.cityId)+" where code="+ str(entity.code))
                c.commit()
            return Airport(entity.code, entity.name, entity.cityId)
        except:
            print("There was an error")
            return None